compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -12025,4 +12043,307 @@ private STNode parseByteArrayContent(SyntaxKind kind) {
+            leadingPipe = parsePipeToken();
+            xmlAtomicNamePatternList.add(leadingPipe);
+            xmlAtomicNamePattern = parseXMLAtomicNamePattern();
+            xmlAtomicNamePatternList.add(xmlAtomicNamePattern);
+            nextToken = peek();
+        }
+
+        return STNodeFactory.createNodeList(xmlAtomicNamePatternList);
+    }
+
+    private boolean isEndOfXMLNamePattern(SyntaxKind tokenKind) {
+        switch (tokenKind) {
+            case IDENTIFIER_TOKEN:
+            case ASTERISK_TOKEN:
+            case COLON_TOKEN:
+                return false;
+            case GT_TOKEN:
+            case EOF_TOKEN:
+            case AT_TOKEN:
+            case DOCUMENTATION_LINE:
+            case CLOSE_BRACE_TOKEN:
+            case SEMICOLON_TOKEN:
+            case PUBLIC_KEYWORD:
+            case PRIVATE_KEYWORD:
+            case FUNCTION_KEYWORD:
+            case RETURNS_KEYWORD:
+            case SERVICE_KEYWORD:
+            case TYPE_KEYWORD:
+            case LISTENER_KEYWORD:
+            case CONST_KEYWORD:
+            case FINAL_KEYWORD:
+            case RESOURCE_KEYWORD:
+                return true;
+            default:
+                return isSimpleType(tokenKind);
+        }
+    }
+
+    /**
+     * Parse xml atomic name pattern.
+     * <p>
+     * <code>
+     * xml-atomic-name-pattern :=
+     *   *
+     *   | identifier
+     *   | xml-namespace-prefix : identifier
+     *   | xml-namespace-prefix : *
+     * </code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseXMLAtomicNamePattern() {
+        STToken token = peek();
+        if (token.kind == SyntaxKind.ASTERISK_TOKEN) {
+            return consume();
+        } else if (token.kind == SyntaxKind.IDENTIFIER_TOKEN) {
+            STNode identifier = consume();
+            return parseXMLAtomicNameIdentifier(identifier);
+        }
+        return STNodeFactory.createEmptyNode();
+    }
+
+    private STNode parseXMLAtomicNameIdentifier(STNode identifier) {
+        STToken token = peek();
+        if (token.kind == SyntaxKind.COLON_TOKEN) {
+            STNode colon = consume();
+            STToken nextToken = peek();
+            if (nextToken.kind == SyntaxKind.IDENTIFIER_TOKEN || nextToken.kind == SyntaxKind.ASTERISK_TOKEN) {
+                STToken endToken = consume();
+                return STNodeFactory.createXMLAtomicNamePatternNode(identifier, colon, endToken);
+            }
+        }
+        return STNodeFactory.createSimpleNameReferenceNode(identifier);
+    }
+
+    /**
+     * Parse xml step expression.
+     * <p>
+     * <code>xml-step-expr := expression xml-step-start xml-step-extend*</code>
+     *
+     * @param lhsExpr Preceding expression of /*, /<, or /**\/< token
+     * @return Parsed node
+     */
+    private STNode parseXMLStepExpression(STNode lhsExpr) {
+        STNode xmlStepStart = parseXMLStepStart();
+        STNode xmlStepExtendList = parseXMLStepExtendList();
```

Comment:
I think we don't need `parseXMLStepExtendList`method at all :) 
Reason is `.<`, `.` and `[` are already handled by parseExpressionRhs method. So xml-step-expression node always have only `lhsExpr` and `xmlStepStart`

Reply:
The problem here is, theres a grammar ambiguity.

Reply:
Then the syntax tree for  `a /* .<*> [e] .b()` xml-step-expr would look like;
```
{
  "kind": "METHOD_CALL",
  "children": [
    {
      "kind": "INDEXED_EXPRESSION",
      "children": [
        {
          "kind": "XML_FILTER_EXPRESSION",
          "children": [
            {
              "kind": "XML_STEP_EXPRESSION",
              "children": [
                {
                  "kind": "IDENTIFIER_TOKEN",
                  "value": "a"
                },
                {
                  "kind": "SLASH_ASTERISK_TOKEN"
                }
              ]
            },
            {
              "kind": "XML_NAME_PATTERN_CHAIN",
              "children": [
                {
                  "kind": "DOT_LT_TOKEN"
                },
                {
                  "kind": "LIST",
                  "children": [
                    {
                      "kind": "ASTERISK_TOKEN"
                    }
                  ]
                },
                {
                  "kind": "GT_TOKEN"
                }
              ]
            }
          ]
        },
        {
          "kind": "OPEN_BRACKET_TOKEN"
        },
        {
          "kind": "LIST",
          "children": [
            {
              "kind": "IDENTIFIER_TOKEN",
              "value": "e"
            }
          ]
        },
        {
          "kind": "CLOSE_BRACKET_TOKEN"
        }
      ]
    },
    {
      "kind": "DOT_TOKEN"
    },
    {
      "kind": "IDENTIFIER_TOKEN",
      "value": "b"
    },
    {
      "kind": "OPEN_PAREN_TOKEN"
    },
    {
      "kind": "LIST",
      "children": []
    },
    {
      "kind": "CLOSE_PAREN_TOKEN"
    }
  ]
}
```
Is it alright? I'm not quite sure.

Reply:
@SupunS plz review the last commit https://github.com/ballerina-platform/ballerina-lang/pull/23581/commits/9c01f069e108ef6ae158daaed3cbfb385371606b for the above change.